ast/term.go

```diff
@@ -2392,6 +2402,54 @@ func filterObject(o Value, filter Value) (Value, error) {
 	}
 }
 
+// NOTE(philipc): The only way to get an ObjectKeyIterator should be
+// from an Object. This ensures that the iterator can have implementation-
+// specific details internally, with no contracts except to the very
+// limited interface.
+type ObjectKeysIterator interface {
+	Next() (*Term, bool)
+	Value() *Term
+	Copy() ObjectKeysIterator
+}
+
+type objectKeysIterator struct {
+	obj     *object
+	numKeys int
+	index   int
+}
+
+func newobjectKeysIterator(o *object) ObjectKeysIterator {
+	return &objectKeysIterator{
+		obj:     o,
+		numKeys: o.Len(),
+		index:   0,
+	}
+}
+
+func (oki *objectKeysIterator) Next() (*Term, bool) {
+	if oki.index == oki.numKeys || oki.numKeys == 0 {
+		return nil, false
+	}
+	oki.index++
+	return oki.obj.sortedKeys()[oki.index-1].key, true
+}
+
+func (oki *objectKeysIterator) Value() *Term {
+	if oki.index == oki.numKeys || oki.numKeys == 0 {
+		return nil
+	}
+	return oki.obj.sortedKeys()[oki.index].key
+}
+
+// Copy returns a deep copy of oki.
```

Comment:
Ah! So the call chain may be helpful for context:
```
(object) Copy()
--> (object) Map()
    --> (object) Iter()
        --> range obj.sortedKeys()
```
That provides the sorted key/value pairs for building up the new object. Since the keys are being handled appropriately at the bottom of the call chain, I didn't feel a need to switch to the `ObjectKeysIterator` style there, although it'd work just fine there if needed.

Reply:
Ah! So the call chain may be helpful for context:
```
(object) Copy()
--> (object) Map()
    --> (object) Iter()
        --> range obj.sortedKeys()
```
That provides the sorted key/value pairs for building up the new object. Since the keys are being handled appropriately at the bottom of the call chain, I didn't feel a need to switch to the `ObjectKeysIterator` style there, although it'd work just fine there if needed.

Reply:
Ah alright. So the new copies end up being sorted already; and we thus don't need to track `numInserted` for them. 

💭 Seems like this is also work we could avoid? OTOH it might not happen too often. I don't know 😅

Reply:
On a related note, if we don't use it, let's remove `Copy() ObjectKeysIterator `.

Reply:
I've removed `Copy()` and `Value()`, since they aren't used anywhere currently, and would be fairly trivial to reimplement if we need them.